### PR TITLE
fix(sql): fix unexpected NPE thrown in query with analytical function and filter

### PR DESCRIPTION
**Pull request by puzpuzpuz**
Also improves close discipline in record cursor factories

### CODE REVIEW

File: core/src/main/java/io/questdb/griffin/engine/orderby/RecordTreeChain.java [hunk 1]
```diff
@@ -376,6 +386,9 @@ public void toTop() {
         }
 
         private void of(RecordCursor base) {
+            if (!isOpen) {
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/orderby/RecordTreeChain.java:389]: nit: redundant if

Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/orderby/RecordTreeChain.java]: Initially I wanted to keep things consistent with what we have in cursor record factories, but I agree that it's redundant. Removed in 492dfb943175c6fc952f26abb83d06524d20c6b5

File: core/src/main/java/io/questdb/griffin/engine/functions/columns/SymbolColumn.java [hunk 2]
```diff
@@ -78,7 +78,13 @@ public void init(SymbolTableSource symbolTableSource, SqlExecutionContext execut
     
     @Override
     public @Nullable StaticSymbolTable getStaticSymbolTable() {
-        return symbolTable instanceof StaticSymbolTable ? (StaticSymbolTable) symbolTable : null;
+        if (symbolTable instanceof StaticSymbolTable) {
+            return (StaticSymbolTable) symbolTable;
+        }
+        if (symbolTable instanceof SymbolFunction) {
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/functions/columns/SymbolColumn.java:84]: If `isSymbolTableStatic() == false`, then `getStaticSymbolTable()` would return a `null` - the behavior of the method is the same with and without the `isSymbolTableStatic()` check. Am I missing anything?

As for the assertion, I've changed it in 38dc3f65f0a972e1a9def7402c1d47549ad1d914 (also added a similar assertion to `MapSymbolColumn`).

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 42 / 44 (95.45%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByRecordCursorFactory.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/MapSymbolColumn.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedRecordCursor.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/FilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilteredRecordCursor.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/columns/SymbolColumn.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/RecordChain.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/eq/EqSymStrFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursor.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedLightRecordCursor.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxReader.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/analytic/CachedAnalyticRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/RecordTreeChain.java|9|9|100.00%|

**Review by ideoma — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 42 / 44 (95.45%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByRecordCursorFactory.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/MapSymbolColumn.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedRecordCursor.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/AsOfJoinRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/FilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilteredRecordCursor.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/columns/SymbolColumn.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/RecordChain.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/eq/EqSymStrFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursor.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/SortedLightRecordCursor.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TxReader.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/analytic/CachedAnalyticRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/RecordTreeChain.java|9|9|100.00%|

**Pull request merged**